Pipeline/controller

UdomkarnBoonyaprasert opened the pull request:
This branch contains the first iteration of the pipeline implementation which should satisfy the first milestone for the HoloPipeline. Installation and usage guide can be found in readme.me in HoloPipeline folder.

UdomkarnBoonyaprasert committed:
Created Python commonents to handle request and data retrieval.
At the moment these components won't do anything. Just a first draft of pipline arch. Current plan is to store the list of all available piplines in a json. (to be discussed later)

UdomkarnBoonyaprasert committed:
Add testing components for json

UdomkarnBoonyaprasert committed:
minor changes for testing. move json into its own folder

UdomkarnBoonyaprasert committed:
Add first draft of pipline architect
available for edit on lucidchart

UdomkarnBoonyaprasert committed:
PIPELINE: update controller

-update json structure
-new argparse in controller
-blank retrieve file

UdomkarnBoonyaprasert committed:
PIPELINE: add HU display tool
run showHU.py to plot HU of a numpy on a chart

UdomkarnBoonyaprasert committed:
PIPELINE: fix numpy2obj
- add kawai's nifti samples
- fix issue with not being able to generate for some nifti
	- need to use showHU to determine the HU to use for each segmented nifti
	- marching cube threshold: changed from int to float
- edit glb outout file naming

UdomkarnBoonyaprasert committed:
PIPELINE: update readme

add new dependencies

UdomkarnBoonyaprasert committed:
PIPELINE: add new components
- add lung segmentation component
- add lung segment to glb pipeline
- add dicom to nifti conversion component
- still some issue with dicom compression in dcm2nifti in some cases
- need to fix output name for dcm2nifti, so that it matches with input name

UdomkarnBoonyaprasert committed:
PIPELINE: update file structure
update file sturcture for pipline and update imports

UdomkarnBoonyaprasert committed:
PIPELINE: delete file

delete outdated files which are not in use

UdomkarnBoonyaprasert committed:
PIPELINE: update readme

dependencies section might be incomplete

UdomkarnBoonyaprasert committed:
PIPELINE: update readme

UdomkarnBoonyaprasert committed:
PIPELINE: add requirements.txt
add req file for pip and remove py file used for experiment

UdomkarnBoonyaprasert committed:
PIPELINE: update readme

add GDCM installation note

UdomkarnBoonyaprasert committed:
PIPELINE: clean up

UdomkarnBoonyaprasert committed:
PIPELINE: delete binary files

UdomkarnBoonyaprasert committed:
PIPELINE: add test files
- add test files
- update gitignore
- bug fixes in components

UdomkarnBoonyaprasert committed:
PIPELINE: update readme
update installation guide and minor fixes to setup.py

boonwj reviewed (changes requested):
Our proposed git workflow is to merge into dev instead of master. Please modify the PR to merge into dev instead. 

*edit: change raising of new PR to modify.

boonwj commented on HoloPipelines/tools/showHU.py:
```diff
@@ -0,0 +1,23 @@
+import os
+import platform
+
+#-----------Unix uses "/", whereas Windows uses "\"-----------
```
Look into using os.path for managing file paths instead of this. It automatically handles the filepaths for the OS the program is running on.

UdomkarnBoonyaprasert replied:
fileHandler.py is now removed

boonwj commented on HoloPipelines/test_pipelines.py:
```diff
@@ -0,0 +1,30 @@
+import pytest
```
Place test files into a test directory. Else the base directory will get too cluttered.

boonwj commented on HoloPipelines/test_pipelineController.py:
```diff
@@ -0,0 +1,37 @@
+import pytest
```
Place test files into a test directory. Else the base directory will get too cluttered.

boonwj commented on HoloPipelines/setup.py:
```diff
@@ -0,0 +1,45 @@
+import urllib.request
```
`setup.py` by python convention is used for package distribution. Perhaps a different naming convention for this file to avoid confusion.

https://stackoverflow.com/questions/1471994/what-is-setup-py

setup.py is a python file, which usually tells you that the module/package you are about to install has been packaged and distributed with Distutils, which is the standard for distributing Python Modules.

UdomkarnBoonyaprasert replied:
setup.py now changed to setupDirectory.py

boonwj commented on HoloPipelines/tools/showHU.py:
```diff
@@ -0,0 +1,23 @@
+
+cwd = os.getcwd() + slash
+
+numpyPath = cwd[:-6] + "numpys" + slash
```
using hardcoded path modification is not safe. This will cause problems when the script is executed from outside this directory or the folder is renamed.

UdomkarnBoonyaprasert replied:
showHU now uses Python's pathlib

boonwj commented on HoloPipelines/README.md:
````diff
@@ -1,41 +1,76 @@
+ cd gdcmbin
+ ccmake ../gdcm
+```
+   [select your configuration] make sure to select *GDCM_BUILD_APPLICATIONS* and *GDCM_BUILD_EXAMPLES*
````
I saw the GDCM python build option is not selected here. Is it correct?

UdomkarnBoonyaprasert replied:
Yes, as the lung segmentation library uses GDCM from the os environment not through a Python import.

boonwj reviewed (changes requested):
Have not ran through everything yet. But please change the way filepaths are handled here. Do not use your own implementation (fileHandler.py). Instead, use os.path.join to craft file paths. 

Please change the way it is done in this project, I will continue reviewing the PR after this.

boonwj commented on HoloPipelines/.gitignore:
```diff
@@ -0,0 +1,13 @@
+__pycache__/
```
The contents in this .gitignore file can be added into the root directory's gitignore. __pycache__ is already ignored there. Just bring the diffs in this to the other file.

boonwj commented on HoloPipelines/README.md:
```diff
@@ -1,41 +1,76 @@
-> python3 dicom2numpy.py
+Some dependencies are not available through pip, they are listed below with their installation instructions
+
+**GDCM (Unix)**
```
GDCM is actually packaged in unix distributions. Have you tried if that works? It would make for a cleaner installation step. 

debian/ubuntu: `sudo apt-get install gdcm`
fedora/redhat: `sudo dnf install gdcm`

UdomkarnBoonyaprasert replied:
I replaced it with 'apt-get' in the readme.

boonwj commented on HoloPipelines/README.md:
````diff
@@ -1,41 +1,76 @@
+```
+pytest
+```
+This however will not display the coverage percentage. (ill write a batch or script for this later. pytest --cov doesnt work for me for some reason)
````
How did the 77% coverage come about?

UdomkarnBoonyaprasert replied:
Through pytest --cov. I have also deleted the coverage number as Niels said it would not make sense as the number can change.

boonwj commented on HoloPipelines/README.md:
````diff
@@ -1,41 +1,76 @@
+```
+Double check to see if installation is successful by running
+```
+ gdcmdump
````
Do mention that the binaries could only be executed in the created `gdcmbin/bin` directory

UdomkarnBoonyaprasert replied:
I've changed it to apt-get so it should be system wide now

UdomkarnBoonyaprasert committed:
PIPELINE: remove fileHandler
- replace fileHandler with Python's pathlib
- fix numpy transpose for nifti with 5D shape in numpy2obj.py
- remove some unused variables

UdomkarnBoonyaprasert committed:
PIPELINE: move test files
- both test files are moved to 'test' directory
- replace os.popen with subprocess.check_output instead
	- add new cwd for subprocess

UdomkarnBoonyaprasert committed:
PIPELINE: update subprocess call
changed os.system() in pipelineController to subprocess.call() with updated path to pipelineList.json
so that pipelineController can be called from any working directory now

nielsboecker commented on HoloPipelines/README.md:
```diff
@@ -1,41 +1,76 @@
+ git clone --branch release git://git.code.sf.net/p/gdcm/gdcm
+ mkdir gdcmbin
+ cd gdcmbin
+ ccmake ../gdcm
```
This is really tedious, includes a dependency to the c compiler, etc ... please do look into using a package manager like Wenjie suggested instead.

UdomkarnBoonyaprasert replied:
readme updated. I deleted the old instruction and replaced it with 2 lines of 'apt-get' for Ubuntu instead.

nielsboecker commented on HoloPipelines/README.md:
```diff
@@ -1,41 +1,76 @@
-https://github.com/AnalyticalGraphicsInc/obj2gltf
-Make sure to have the latest version of Node.js 
+## Testing:
+Testing has a 77% code coverage. Testing can be done by using pytest:
```
You don't want to write this number into the README. Think about it: It will be obsolete as soon as you commit a couple of lines to the project, or delete, or add a test...

UdomkarnBoonyaprasert replied:
readme updated.

boonwj commented on HoloPipelines/test/test_pipelines.py:
```diff
@@ -0,0 +1,35 @@
+	output = subpro("python pipelineController.py p1 -p 1103_3_glm.nii 30 0", cwd=newCwd, shell=True).decode("utf-8")
+	assert "nifti2glb: done" in str(output)
+	assert os.path.isfile(glbPath.joinpath("1103_3_glm.glb"))
+
```
Tests here does not perform cleanups. So tests that checks for an output file's existence is not a reliable check.

UdomkarnBoonyaprasert replied:
I've added fixture to the tests. So it would do the cleanup after each test case

boonwj commented on HoloPipelines/pipelines/pipelineList.json:
```diff
@@ -0,0 +1 @@
+{"s0": [{"name": "sample0", "src": "pipelines/sample0.py", "info": "A pipline for testing. It will printout 'Hello World! from 0'", "addDate": "26/06/2019", "modDate": "26/06/2019", "param": "0"}], "s1": [{"name": "sample1", "src": "pipelines/sample1.py", "info": "A pipline for testing. It will printout 'Hello World! from 1'", "addDate": "26/06/2019", "modDate": "26/06/2019", "param": "0"}], "i0" ...
```
This is not readable for a configuration file. Do format it into a human-readable format.

UdomkarnBoonyaprasert replied:
reformatted the json

boonwj commented on HoloPipelines/test/test_pipelineController.py:
```diff
@@ -0,0 +1,34 @@
+from subprocess import check_output as subpro
+
+newCwd = str(pathlib.Path(str(os.path.dirname(os.path.realpath(__file__)))).parent)
+
```
All tests here are highly dependent on the current configurations being fed into pipelineController. So any changes in the configuration may break the tests. 

Look into a different way of implementing the tests so that it can be more robust. (e.g. specify test configurations to be loaded)

UdomkarnBoonyaprasert replied:
I've implemented the code so now it can load any specific pipeline config file(json)

boonwj commented on HoloPipelines/test/test_pipelines.py:
```diff
@@ -0,0 +1,35 @@
+glbPath = outputPath.joinpath("GLB")
+
+newCwd = str(pathlib.Path(str(os.path.dirname(os.path.realpath(__file__)))).parent)
+
```
Similar comment to the test_pipelineController. All tests are highly dependent on current configurations. Changes made to the configurations will fail the tests

UdomkarnBoonyaprasert replied:
Changes made to the pipelines and components now allows the user to specify their own input and output path

boonwj commented on HoloPipelines/pipelines/nifti2glb.py:
```diff
@@ -0,0 +1,38 @@
+	print("nifti2glb: done")
+
+if __name__ == '__main__':
+	import sys
```
why are you importing sys here?

UdomkarnBoonyaprasert replied:
removed

boonwj commented on HoloPipelines/pipelines/nifti2obj.py:
```diff
@@ -0,0 +1,29 @@
+
+argCount = len(sys.argv)
+
+if argCount != 4:#need changing
```
what does this need to be changed to?

UdomkarnBoonyaprasert replied:
Moved to pipelineController.py

boonwj commented on HoloPipelines/pipelines/nifti2glb.py:
```diff
@@ -0,0 +1,38 @@
+	if argCount == 4:
+		outputName = sys.argv[3]
+
+	tempNii = niftiData.replace(".nii.gz", "").replace(".nii", "")
```
this segment looks to be repeating code in main(). When don't you pass the arguments to main() instead?

UdomkarnBoonyaprasert replied:
All pipelines now call code from its own main()

boonwj commented on HoloPipelines/pipelines/lungDicom2glb.py:
```diff
@@ -0,0 +1,30 @@
+		"lungDicom2glb: invalid number of argument"
+		exit()
+	fname = str(sys.argv[1])
+	dcm2nifti.main(str(fname))
```
Code here is repeated from main(), just call main() instead.

UdomkarnBoonyaprasert replied:
All pipelines now call code from its own main()

boonwj commented on HoloPipelines/pipelines/components/obj2gltfWrapper.py:
```diff
@@ -0,0 +1,22 @@
+def main(fname="-h", delObj=False):
+	success = subprocess.run(["obj2gltf", "-i", str(objPath.joinpath(fname)), "-b"])
+	if success:
+		success = subprocess.run(["mv", str(objPath.joinpath(fname.replace(".obj", ".glb"))), str(glbPath)])
```
Don't use subprocess to call basic file operation tools like move, rename and delete. The python script would then be effectively a bash script. Use basic functionality provided by the os library (os.rename, os.remove)

UdomkarnBoonyaprasert replied:
All basic file operations now use os or shutil

boonwj commented on HoloPipelines/pipelines/components/numpy2obj.py:
```diff
@@ -0,0 +1,77 @@
+	return verts, faces, norm
+
+def makeObj(fPath, thisThreshold, objOutput):
+	if str(type(fPath)) == "<class 'numpy.ndarray'>":
```
Please isinstance() for this

boonwj replied:
isinstance(fPath, np.ndarray)

UdomkarnBoonyaprasert replied:
done

boonwj commented on HoloPipelines/pipelines/components/lungSegment/segment_airway.py:
```diff
@@ -0,0 +1,46 @@
+import numpy as np
+import nibabel as nib
+from scipy import ndimage
+#import fistme
```
dude pap wtf is this? lol

UdomkarnBoonyaprasert replied:
removed xd

boonwj commented on HoloPipelines/pipelines/lungDicom2glb.py:
```diff
@@ -0,0 +1,30 @@
+def main(fname):
+	dcm2nifti.main(str(fname))
+	lungSegment.main(lower(str(fname)) + ".nii.gz")
+	sucess = subprocess.run(["mv", str(cwd.joinpath("pipelines", "components", "lungSegment", "result", "sample_lung.nii.gz")), str(niftiPath)])
```
Try not to use subprocess for basic linux operations

UdomkarnBoonyaprasert replied:
Replaced to os.move

boonwj commented on HoloPipelines/pipelines/lungDicom2glb.py:
```diff
@@ -0,0 +1,30 @@
+#THERE IS THIS PROBLEM THAT DCM2NIFTI SAVES USING STUFF FROM DICOM META
+#TODO this pipeline will always save as sample_lung.glb as it gets its nii from lungSegment
```
what's there left todo here

UdomkarnBoonyaprasert replied:
Comments removed

boonwj commented on HoloPipelines/pipelines/nifti2glb.py:
```diff
@@ -0,0 +1,38 @@
+
+	if argCount != 3 and argCount != 4:
+		print("Invalid number of argument")
+		exit()
```
sys.exit("Invalid number of argument")

UdomkarnBoonyaprasert replied:
done

UdomkarnBoonyaprasert committed:
PIPELINE: cleanup
- pipelineController now can accept pipeline config file outside the default dir by using the flag -c (--config)
- add config file for testing
- move sample pipelines to testing
- add setup/teardown for pytest
- minor changes to test files
- update code structure of pipeline files
- comment clean up

UdomkarnBoonyaprasert committed:
PIPELINE: update readme

boonwj commented on HoloPipelines/pipelines/components/lungSegment/demo_ct_lung_segment.ipynb:
```diff
@@ -0,0 +1,516 @@
+{
```
This is a not a component and it's more for documentation. Perhaps put this file into a separate documentation directory.

UdomkarnBoonyaprasert replied:
removed

boonwj commented on HoloPipelines/temptest.py:
```diff
@@ -0,0 +1,57 @@
+import os
```
this file is weird as hell dude lol

UdomkarnBoonyaprasert replied:
removed

boonwj commented on HoloPipelines/test/test_pipelines.py:
```diff
@@ -0,0 +1,95 @@
+def testSetup():
+	print('Beginning dicom sample download...')
+
+	if not os.path.exists(str(niftiPath.joinpath("1103_3_glm.nii"))) and os.path.exists(str(dicomPath.joinpath("3_Axial_CE"))):
```
logic seems weird

UdomkarnBoonyaprasert replied:
Code now check for each files separately

boonwj commented on HoloPipelines/test/test_pipelines.py:
```diff
@@ -0,0 +1,95 @@
+
+		print("setup: done")
+
+	if os.path.exists(str(glbPath.joinpath("3_Axial_CE.glb"))):#have them for each test case?
```
duplicated code here

boonwj replied:
oh just realised this was clean up during setup. The second call is for cleanup during teardown. You might want to refactor the code to something like the following

```
files = ["file1", "file2"]
for junk_files in files:
    <clean up clode logic>
```

boonwj commented on HoloPipelines/test/test_pipelineController.py:
```diff
@@ -0,0 +1,51 @@
+newCwd = str(pathlib.Path(str(os.path.dirname(os.path.realpath(__file__)))).parent)
+
+def test_main_s0_0():
+	output = checkOutput('python pipelineController.py -c test/testList.json s0', cwd=newCwd, shell=True).decode("utf-8")
```
do not mix checkOutput and subprocess.check_output. just use 1

UdomkarnBoonyaprasert replied:
done

boonwj commented on HoloPipelines/test/sample1.py:
```diff
@@ -0,0 +1 @@
+print("Hello World! from 01")
```
this file does not seem to be used. Remove it

UdomkarnBoonyaprasert replied:
removed

boonwj commented on HoloPipelines/test/testList.json:
```diff
@@ -0,0 +1,92 @@
+         "param":"0"
+      }
+   ],
+   "s1":[
```
this test configuration is not used

UdomkarnBoonyaprasert replied:
unused configs now removed

boonwj commented on HoloPipelines/pipelines/lungDicom2glb.py:
```diff
@@ -0,0 +1,37 @@
+
+	main(fname)
+
+	'''dcm2nifti.main(str(fname))
```
remove unnecessary commented code

UdomkarnBoonyaprasert replied:
removed

boonwj commented on HoloPipelines/pipelines/components/obj2gltfWrapper.py:
```diff
@@ -0,0 +1,41 @@
+		###################################################################################################
+	else:
+		print("obj2gltf: conversion failed")
+		exit()
```
sys.exit()

UdomkarnBoonyaprasert replied:
done

boonwj commented on HoloPipelines/pipelines/components/obj2gltfWrapper.py:
```diff
@@ -0,0 +1,41 @@
+import subprocess
+import pathlib
+from shutil import move as move
```
`from shutil import move as move` is no different from `from shutil import move`

UdomkarnBoonyaprasert replied:
done

boonwj commented on HoloPipelines/pipelines/components/obj2gltfWrapper.py:
```diff
@@ -0,0 +1,41 @@
+import subprocess
+import pathlib
+from shutil import move as move
+from os import remove as remove
```
same as above

UdomkarnBoonyaprasert replied:
done

boonwj commented on HoloPipelines/pipelines/components/obj2gltfWrapper.py:
```diff
@@ -0,0 +1,41 @@
+import pathlib
+from shutil import move as move
+from os import remove as remove
+from os import rename as rename
```
same as above

UdomkarnBoonyaprasert replied:
yes sir

boonwj commented on HoloPipelines/pipelines/components/obj2gltfWrapper.py:
```diff
@@ -0,0 +1,41 @@
+		move(str(objPath.joinpath(fname.replace(".obj", ".glb"))), newGlbPath)#one line to solve all prob?
+		if delObj:
+			remove(str(objPath.joinpath(fname)))
+		'''if "__temp__" in str(fname):
```
unnecessary comments

UdomkarnBoonyaprasert replied:
removed

boonwj commented on HoloPipelines/pipelines/dicom2glb.py:
```diff
@@ -0,0 +1,19 @@
+
+dicomPath = pathlib.Path.cwd().joinpath("medicalScans", "dicom")
+
+def main(dicomData, threshold):
```
naming convention is not clear what dicomData actually is

UdomkarnBoonyaprasert replied:
Renamed to dicomFolderPath

boonwj commented on HoloPipelines/pipelines/lungDicom2glb.py:
```diff
@@ -0,0 +1,37 @@
+niftiPath = cwd.joinpath("medicalScans", "nifti")
+
+def main(fname):
+	dcm2nifti.main(str(fname))
```
can you be more consistent with what you trust as a string

UdomkarnBoonyaprasert replied:
Should be a lot more consistent now

boonwj commented on HoloPipelines/pipelines/lungDicom2glb.py:
```diff
@@ -0,0 +1,37 @@
+	os.remove(str(niftiPath.joinpath("sample_lung.nii.gz")))
+	print("lungDicom2glb: done")
+
+	'''dcm2nifti.main(str(fname))
```
unnecessary comments can be removed

UdomkarnBoonyaprasert replied:
removed

boonwj commented on HoloPipelines/pipelines/nifti2glb.py:
```diff
@@ -0,0 +1,30 @@
+niftiPath = pathlib.Path.cwd().joinpath("medicalScans", "nifti")
+
+def main(fname, threshold, outputName=""):
+	tempNii = fname.replace(".nii.gz", "").replace(".nii", "")
```
allow method to specify input filepaths and output filepaths. So you don't have to guess

UdomkarnBoonyaprasert replied:
Done for all pipelines and components

boonwj commented on HoloPipelines/test/test_pipelineController.py:
```diff
@@ -0,0 +1,51 @@
+	except subprocess.CalledProcessError as temp:
+		output = temp.returncode
+	assert "1" in str(output)
+	'''output = checkOutput('python pipelineController.py -c test/testList.json s3', cwd=newCwd, shell=True).decode("utf-8")
```
unnecessary comment

UdomkarnBoonyaprasert replied:
removed

boonwj reviewed (changes requested):
Shit i didn't click this. wtf dude sorry.

UdomkarnBoonyaprasert committed:
PIPELINE: PR clean up
- change config file structure
- rename pipeline and components
- re-structure pipelines
- change pipelines and components arguments
- delete unused comments

UdomkarnBoonyaprasert committed:
PIPELINE: update readme
update dependencies in readme

UdomkarnBoonyaprasert committed:
PIPELINE: delete comments
delete unsed comments

boonwj commented on HoloPipelines/README.md:
````diff
@@ -11,44 +11,32 @@ pip install -r requirements.txt
- make
-```
 Double check to see if installation is successful by running
 ```
````
`gdcmdump` or `gdcmconv`

boonwj commented on HoloPipelines/README.md:
```diff
@@ -1,41 +1,76 @@
-# dicom2mesh
-A simple Python script which utilises marching cubes algorithm(skimage library) to generate 3D mesh from Dicom/Nifti image stack, which then can be output to .obj file(can be previewed using 3D viewer from microsoft store).
+# PIPELINE
```
Name to be more explicit. i.e. HoloPipelines. Suggestions @nbckr?

nielsboecker replied:
Yes, HoloPipelines, to be coherent within our terminology.

UdomkarnBoonyaprasert replied:
Changed to HoloPipelines

boonwj commented on HoloPipelines/README.md:
````diff
@@ -1,41 +1,76 @@
+```
+gdcmconv
+```
+If command is not found then please make sure to add to PATH
````
how do you add to path?

UdomkarnBoonyaprasert replied:
Instruction added

boonwj commented on HoloPipelines/README.md:
````diff
@@ -1,41 +1,76 @@
+```
+python pipelineController.py -h
+```
+- this script has a built in help
````
Put the output of the print here.

boonwj commented on HoloPipelines/README.md:
````diff
@@ -1,41 +1,76 @@
+pytest --cov
+```
 
 ### Refs:
````
don't think the refs here are necessary for the program

UdomkarnBoonyaprasert replied:
Removed

boonwj commented on HoloPipelines/pipelineController.py:
```diff
@@ -0,0 +1,72 @@
+		if args.ls:
+			print(json.dumps(lsPipe, indent=4, sort_keys=False))
+			sys.exit()
+		elif len(args.info) > 0:
```
if

UdomkarnBoonyaprasert replied:
Replaced to if

boonwj commented on HoloPipelines/pipelineController.py:
```diff
@@ -0,0 +1,72 @@
+		#check if pipeline exist
+		if args.pipelineID not in lsPipe:
+			sys.exit("pipelineController: no pipeline with such ID")
+		if len(args.param) != int(lsPipe[args.pipelineID][0]['param']):
```
Rethink the need for hardcoded 0 for configurations

UdomkarnBoonyaprasert replied:
config file structure updated

boonwj commented on HoloPipelines/pipelines/components/compDicom2numpy.py:
```diff
@@ -0,0 +1,87 @@
+	
+	return image, new_spacing
+
+def main(dicomPath, outputNumpy=False, outputNumpyPath=""):
```
If this is the only reason for the `numpy` directory It should be not exposed to the users. You can also remove the folder creations for `numpy` directory

boonwj replied:
This is also a component that will not be configured to run in the pipeline configurations

UdomkarnBoonyaprasert replied:
The option to output numpy has been removed along with numpy directory creation in pipelineController.py

boonwj commented on HoloPipelines/pipelines/components/compNifti2numpy.py:
```diff
@@ -0,0 +1,48 @@
+	
+	return image, new_spacing
+
+def main(inputNiftiPath, outputNumpyPath=""):
```
Keep usage consistent with `compDicom2numpy`.

UdomkarnBoonyaprasert replied:
updated

boonwj commented on HoloPipelines/pipelines/components/compNifti2numpy.py:
```diff
@@ -0,0 +1,48 @@
+
+numpyPath = cwd.joinpath("numpy")
+
+fName = ""
```
????

UdomkarnBoonyaprasert replied:
removed

boonwj commented on HoloPipelines/pipelines/components/compNifti2numpy.py:
```diff
@@ -0,0 +1,48 @@
+numpyPath = cwd.joinpath("numpy")
+
+fName = ""
+tempPath = ""
```
?????

UdomkarnBoonyaprasert replied:
removed

boonwj commented on HoloPipelines/pipelines/components/compNumpy2obj.py:
```diff
@@ -0,0 +1,70 @@
+from sklearn.cluster import KMeans
+import pathlib
+import sys
+
```
#https://github.com/nipy/nibabel/issues/626

boonwj commented on HoloPipelines/pipelines/components/compNumpy2obj.py:
```diff
@@ -0,0 +1,70 @@
+cwd = pathlib.Path.cwd()
+
+objPath = cwd.joinpath("output", "OBJ")
+numpyPath = cwd.joinpath("numpy")
```
???

UdomkarnBoonyaprasert replied:
removed

boonwj commented on HoloPipelines/pipelines/components/compNumpy2obj.py:
```diff
@@ -0,0 +1,70 @@
+
+cwd = pathlib.Path.cwd()
+
+objPath = cwd.joinpath("output", "OBJ")
```
???

UdomkarnBoonyaprasert replied:
removed

boonwj commented on HoloPipelines/pipelines/components/compNumpy2obj.py:
```diff
@@ -0,0 +1,70 @@
+		try:
+			numpyData = np.load(str(pathlib.Path(inputNumpy)))
+		except:
+			sys.exit("numpy2obj: error occured while loading numpy. please make sure the path to numpy is correct.")
```
*P*lease

UdomkarnBoonyaprasert replied:
updated

boonwj commented on HoloPipelines/pipelines/components/compNumpy2obj.py:
```diff
@@ -0,0 +1,70 @@
+def main(inputData, mainThreshold, outputPath):
+	generateObj(inputData, mainThreshold, outputPath)
+	return outputPath
+	print ("numpy2obj: done")
```
this won't run right

UdomkarnBoonyaprasert replied:
updated

boonwj commented on HoloPipelines/pipelines/components/compObj2gltfWrapper.py:
```diff
@@ -0,0 +1,33 @@
+
+success = True
+cwd = pathlib.Path.cwd()
+objPath = cwd.joinpath("output", "OBJ")
```
???

UdomkarnBoonyaprasert replied:
Unused codes removed

boonwj commented on HoloPipelines/pipelines/components/compObj2gltfWrapper.py:
```diff
@@ -0,0 +1,33 @@
+success = True
+cwd = pathlib.Path.cwd()
+objPath = cwd.joinpath("output", "OBJ")
+glbPath = cwd.joinpath("output", "GLB")
```
???

UdomkarnBoonyaprasert replied:
Unused codes removed

boonwj commented on HoloPipelines/pipelines/components/compObj2gltfWrapper.py:
```diff
@@ -0,0 +1,33 @@
+glbPath = cwd.joinpath("output", "GLB")
+
+def main(inputObjPath, outputGlbPath, deleteOriginalObj=False, compressGlb=False):
+	success = subprocess.run(["obj2gltf", "-i", str(pathlib.Path(inputObjPath)), "-b"])#TODO check return
```
Do you still need the #todo flag

UdomkarnBoonyaprasert replied:
Unused comment removed

boonwj commented on HoloPipelines/pipelines/components/compObj2gltfWrapper.py:
```diff
@@ -0,0 +1,33 @@
+objPath = cwd.joinpath("output", "OBJ")
+glbPath = cwd.joinpath("output", "GLB")
+
+def main(inputObjPath, outputGlbPath, deleteOriginalObj=False, compressGlb=False):
```
we might want to rethink the default for keeping obj if it's never gonna be used

UdomkarnBoonyaprasert replied:
Default changed to deleteOriginalObj=Ture

boonwj commented on HoloPipelines/pipelines/components/compObj2gltfWrapper.py:
```diff
@@ -0,0 +1,33 @@
+
+		#Draco compression. note that draco compresssion in viewers may not be common
+		if compressGlb:
+			subprocess.call("gltf-pipeline -i " + outputGlbPath + " -o " + outputGlbPath + " -d", cwd=newCwd, shell=True)
```
you used list input for subproces call previously.  Why is it changed to shell here? The best practice is to not use shell.

UdomkarnBoonyaprasert replied:
Changed to subprocess.run() without shell=True

boonwj commented on HoloPipelines/pipelines/components/lungSegment/README.md:
```diff
@@ -0,0 +1,30 @@
+* scipy
+* scikit-image
+
+## Usage
```
either update the usage to how we can use it for the pipelines or if not remove it

UdomkarnBoonyaprasert replied:
README.md removed

boonwj commented on HoloPipelines/pipelineController.py:
```diff
@@ -0,0 +1,72 @@
+			sys.exit("pipelineController: invalid number of param [expected: " + str(lsPipe[args.pipelineID][0]['param']) + ", got: " + str(len(args.param)) + "]")
+		#start pipeline
+		print("starting pipeline "+args.pipelineID+"...")
+		call('python ' + lsPipe[args.pipelineID][0]['src'] + " " +" ".join(args.param), cwd=newCwd, shell=True)
```
Use the non shell way of running calls

UdomkarnBoonyaprasert replied:
Code updated to not used shell=True

boonwj commented on HoloPipelines/pipelineController.py:
```diff
@@ -0,0 +1,72 @@
+import os
+import sys
+import argparse
+from subprocess import call
```
Be more consistent with subprocess call usage. Just use subprocess.call()

UdomkarnBoonyaprasert replied:
Updateed to subprocess.run()

boonwj commented on HoloPipelines/pipelines/components/compDicom2numpy.py:
```diff
@@ -0,0 +1,87 @@
+import time
+
+cwd = pathlib.Path.cwd()
+numpyPath = cwd.joinpath("numpy")
```
not necessary when script takes in full path.

UdomkarnBoonyaprasert replied:
Unused codes removed

boonwj commented on HoloPipelines/pipelines/components/compObj2gltfWrapper.py:
```diff
@@ -0,0 +1,33 @@
+import subprocess
```
gltf and glb are 2 different things. So it might be more explicit to indiciate in the filename what you are returning. `compObj2GlbWrapper.py
`

UdomkarnBoonyaprasert replied:
Components renamed to compObj2GlbWrapper.py

boonwj commented on HoloPipelines/pipelines/components/compObj2gltfWrapper.py:
```diff
@@ -0,0 +1,33 @@
+		move(str(pathlib.Path(inputObjPath)).replace(".obj", ".glb"), outputGlbPath)
+		if deleteOriginalObj:
+			os.remove(str(pathlib.Path(inputObjPath)))
+		print("obj2gltf: conversion complete")
```
this might be obj2glb

UdomkarnBoonyaprasert replied:
Renamed

boonwj commented on HoloPipelines/pipelines/components/compObj2gltfWrapper.py:
```diff
@@ -0,0 +1,33 @@
+		#Draco compression. note that draco compresssion in viewers may not be common
+		if compressGlb:
+			subprocess.call("gltf-pipeline -i " + outputGlbPath + " -o " + outputGlbPath + " -d", cwd=newCwd, shell=True)
+			print("obj2gltf: Draco compression finished")
```
obj2glb

UdomkarnBoonyaprasert replied:
Renamed

boonwj commented on HoloPipelines/pipelines/components/compObj2gltfWrapper.py:
```diff
@@ -0,0 +1,33 @@
+			subprocess.call("gltf-pipeline -i " + outputGlbPath + " -o " + outputGlbPath + " -d", cwd=newCwd, shell=True)
+			print("obj2gltf: Draco compression finished")
+	else:
+		sys.exit("obj2gltf: conversion failed")
```
obj2glb

UdomkarnBoonyaprasert replied:
Renamed

boonwj commented on HoloPipelines/pipelines/components/compObj2gltfWrapper.py:
```diff
@@ -0,0 +1,33 @@
+	else:
+		sys.exit("obj2gltf: conversion failed")
+	return outputGlbPath
+	print("obj2gltf: done")
```
obj2glb

UdomkarnBoonyaprasert replied:
Renamed

boonwj commented on HoloPipelines/pipelines/components/lungSegment/main.py:
```diff
@@ -0,0 +1,85 @@
+import glob
+import sys
+
+niftiPath = pathlib.Path.cwd().joinpath("medicalScans", "nifti")
```
???

UdomkarnBoonyaprasert replied:
Unused codes removed

boonwj commented on HoloPipelines/pipelines/components/lungSegment/main.py:
```diff
@@ -0,0 +1,85 @@
+	from components.lungSegment.segment_airway import segment_airway
+
+	params = utils.define_parameter()
+
```
If you can handle either directory or file. Then be more explicit with your checks.

os.path.isdir() or os.path.isfile()

UdomkarnBoonyaprasert replied:
Updated to os.path.isdir() to check for directory

boonwj commented on HoloPipelines/pipelines/components/lungSegment/main.py:
```diff
@@ -0,0 +1,85 @@
+
+	Mlung, Maw = segment_airway(params, I, I_affine, Mlung)
+
+	return str(pathlib.Path.cwd().joinpath("pipelines", "components", "lungSegment", "result", "sample_lung.nii.gz"))
```
Same filename may lead to concurrency issues when we ever do that. Might need to prefix or suffix the filename with something. @nbckr inputs?

boonwj commented on HoloPipelines/pipelines/components/lungSegment/main.py:
```diff
@@ -0,0 +1,85 @@
+
+
+if __name__ == '__main__':
+	try:
```
To keep to the same performance as other components. Might be better to keep the usage similar (return error on standalone executions)

UdomkarnBoonyaprasert replied:
Now calls main()

boonwj commented on HoloPipelines/pipelines/components/lungSegment/segment_airway.py:
```diff
@@ -0,0 +1,45 @@
+	Mawtmp = ndimage.binary_dilation(Maw, structure = struct_l, iterations = 1)
+	Mawtmp = np.int8(Mawtmp)
+	Mlung[Maw > 0] = 0
+	nib.Nifti1Image(Maw,I_affine).to_filename('pipelines/components/lungSegment/result/sample_aw.nii.gz')
```
Create a function argument that changes the output path. So you won't be bound to this folder structure.

UdomkarnBoonyaprasert replied:
lungSegment now takes output folder path as one of its arguments

boonwj commented on HoloPipelines/pipelines/components/lungSegment/segment_airway.py:
```diff
@@ -0,0 +1,45 @@
+except:
+	from components.lungSegment.utils import *
+
+def segment_airway(params, I, I_affine, Mlung):
```
Create a function argument that changes the output path. So you won't be bound to this folder structure.

UdomkarnBoonyaprasert replied:
lungSegment now takes output folder path as one of its arguments

boonwj commented on HoloPipelines/pipelines/components/lungSegment/segment_lung.py:
```diff
@@ -0,0 +1,63 @@
+from skimage import measure
+from scipy import stats
+
+def segment_lung(params, I, I_affine):
```
Create a function argument that changes the output path. So you won't be bound to this folder structure.

UdomkarnBoonyaprasert replied:
lungSegment now takes output folder path as one of its arguments

boonwj commented on HoloPipelines/pipelines/components/lungSegment/segment_lung.py:
```diff
@@ -0,0 +1,63 @@
+	M     = ndimage.binary_fill_holes(M)
+	Mlung = np.int8(M)
+	#nib.Nifti1Image(Mlung,I_affine).to_filename('./result/sample_lungaw.nii.gz')
+	nib.Nifti1Image(Mlung,I_affine).to_filename('pipelines/components/lungSegment/result/sample_lungaw.nii.gz')
```
Create a function argument that changes the output path. So you won't be bound to this folder structure.

UdomkarnBoonyaprasert replied:
lungSegment now takes output folder path as one of its arguments

boonwj commented on HoloPipelines/pipelines/pipelineDicom2glb.py:
```diff
@@ -0,0 +1,14 @@
+import pathlib
+import sys
+
+def main(dicomFolderPath, threshold, outputGlbPath):
```
function(in, out, config)

boonwj replied:
`main("input_path", threshold=99, outputGlbPath="output_path")`
if changed
`main("input_path", "output_path", threshold=99)`

UdomkarnBoonyaprasert replied:
Arguments re-ordered

boonwj commented on HoloPipelines/pipelines/pipelineDicom2glb.py:
```diff
@@ -0,0 +1,14 @@
+
+def main(dicomFolderPath, threshold, outputGlbPath):
+	generatedNumpyList = compDicom2numpy.main(str(pathlib.Path(dicomFolderPath)))
+	generatedObjPath = compNumpy2obj.main(generatedNumpyList, threshold, str(pathlib.Path.cwd().joinpath("output", "OBJ").joinpath(str(pathlib.PurePath(dicomFolderPath).parts[-1]))) + ".obj")
```
dicomFolderPath = input/dicom_stuff
generatedObjPath = cwd/output/OBJ/dicom_stuff.obj

boonwj commented on HoloPipelines/pipelines/pipelineDicom2glb.py:
```diff
@@ -0,0 +1,14 @@
+def main(dicomFolderPath, threshold, outputGlbPath):
+	generatedNumpyList = compDicom2numpy.main(str(pathlib.Path(dicomFolderPath)))
+	generatedObjPath = compNumpy2obj.main(generatedNumpyList, threshold, str(pathlib.Path.cwd().joinpath("output", "OBJ").joinpath(str(pathlib.PurePath(dicomFolderPath).parts[-1]))) + ".obj")
+	generatedGlbPath = compObj2gltfWrapper.main(generatedObjPath, outputGlbPath, True, False)
```
Instead of just True, False. You can call it as flag=True, flag2=False

UdomkarnBoonyaprasert replied:
Flags added

boonwj commented on HoloPipelines/pipelines/pipelineList.json:
```diff
@@ -0,0 +1,42 @@
+{
+   "p0":[
```
Rethink the need for [] in your configuration

UdomkarnBoonyaprasert replied:
Config files structure updated

boonwj commented on HoloPipelines/pipelines/pipelineLungDicom2glb.py:
```diff
@@ -0,0 +1,20 @@
+	generatedSegmentedLungsNiftiPath = lungSegment(generatedNiftiPath)
+	generatedNumpyList = compNifti2numpy.main(str(pathlib.Path(generatedSegmentedLungsNiftiPath)))
+	generatedObjPath = compNumpy2obj.main(generatedNumpyList, 0.5, str(pathlib.Path.cwd().joinpath("output", "OBJ", "nifti2glb_tempObj.obj")))
+	generatedGlbPath = compObj2gltfWrapper.main(generatedObjPath, str(pathlib.Path(outputGlbPath)), True, False)
```
Use Flag=True, Flag2=False for clarity.

UdomkarnBoonyaprasert replied:
Flags have been added

boonwj commented on HoloPipelines/pipelines/pipelineLungDicom2glb.py:
```diff
@@ -0,0 +1,20 @@
+#NOTE: lungSegment will always output as sample_lung.nii.gz etc
```
This needs to change once the function for lung segmentation is changed.

UdomkarnBoonyaprasert replied:
Comments removed

boonwj commented on HoloPipelines/pipelines/pipelineNifti2glb.py:
```diff
@@ -0,0 +1,14 @@
+def main(inputNiftiPath, threshold, outputGlbPath):
+	generatedNumpyList = compNifti2numpy.main(str(pathlib.Path(inputNiftiPath)))
+	generatedObjPath = compNumpy2obj.main(generatedNumpyList, threshold, str(pathlib.Path.cwd().joinpath("output", "OBJ", "nifti2glb_tempObj.obj")))
+	generatedGlbPath = compObj2gltfWrapper.main(generatedObjPath, str(pathlib.Path(outputGlbPath)), True, False)
```
Flag1=, Flag2=

UdomkarnBoonyaprasert replied:
Flags added

boonwj commented on HoloPipelines/pipelines/pipelineNifti2obj.py:
```diff
@@ -0,0 +1,14 @@
+import sys
+import numpy as np
+
+def main(inputNiftiPath, threshold, outputObjPath, flipNpy=False):
```
func(in, out, other stuff)

UdomkarnBoonyaprasert replied:
Arguments reordered

boonwj commented on HoloPipelines/setupExample.py:
```diff
@@ -0,0 +1,49 @@
+	os.mkdir("output")
+	os.mkdir("output/OBJ")
+	os.mkdir("output/GLB")
+if not os.path.exists("pipelines/components/lungSegment/result"):
```
when your lung segmentation implementation changes. See if you still need this.

UdomkarnBoonyaprasert replied:
Code removed

boonwj commented on HoloPipelines/setupExample.py:
```diff
@@ -0,0 +1,49 @@
+	os.mkdir("medicalScans")
+	os.mkdir("medicalScans/dicom")
+	os.mkdir("medicalScans/nifti")
+if not os.path.exists("numpy"):
```
do you need this?

UdomkarnBoonyaprasert replied:
Code removed

boonwj commented on HoloPipelines/test/testList.json:
```diff
@@ -0,0 +1,52 @@
+{
+   "s0":[
```
if you do remove [] from json. Make sure your test passes too.

UdomkarnBoonyaprasert replied:
Test config file also updated

boonwj commented on HoloPipelines/test/test_pipelineController.py:
```diff
@@ -0,0 +1,50 @@
+newCwd = str(pathlib.Path(str(os.path.dirname(os.path.realpath(__file__)))).parent)
+
+def test_main_s0_0():
+	output = subprocess.check_output('python pipelineController.py -c test/testList.json s0', cwd=newCwd, shell=True).decode("utf-8")
```
use the non-shell version of the calls. (https://docs.python.org/3/library/subprocess.html)

UdomkarnBoonyaprasert replied:
Updated

boonwj commented on HoloPipelines/setupExample.py:
```diff
@@ -0,0 +1,49 @@
+import urllib.request
```
You might not need this any more if your test_pipelines.py file already does the setup for you.

UdomkarnBoonyaprasert replied:
setupExample.py removed

boonwj commented on HoloPipelines/test/test_pipelines.py:
```diff
@@ -0,0 +1,85 @@
+
+	remove3Dmodels()
+
+	setupEndsHere = True
```
you don't really need this. Replace it with a comment or something.

UdomkarnBoonyaprasert replied:
updated

boonwj reviewed (changes requested):
Awesome man I can finally see the light.

UdomkarnBoonyaprasert committed:
PIPELINE: PR cleanup
-replaced subprocess.call() to subprocess.run() and removed shell=True
-remove unused code
-removed unused files

boonwj commented on HoloPipelines/pipelines/components/compObj2glbWrapper.py:
```diff
@@ -7,27 +7,24 @@
-	success = subprocess.run(["obj2gltf", "-i", str(pathlib.Path(inputObjPath)), "-b"])#TODO check return
+def main(inputObjPath, outputGlbPath, deleteOriginalObj=True, compressGlb=False):
+	success = subprocess.run(["obj2gltf", "-i", str(pathlib.Path(inputObjPath)), "-b"])
 	if success:
```
This statement will always return `true` as it checks for the existence of the `CompletedProcess` object. Instead, you need to check the `returncode` within the `CompletedProcess` object

UdomkarnBoonyaprasert replied:
Changed to `if success.returncode == 0`

boonwj commented on HoloPipelines/pipelines/components/compObj2glbWrapper.py:
```diff
@@ -7,27 +7,24 @@
 		if compressGlb:
-			subprocess.call("gltf-pipeline -i " + outputGlbPath + " -o " + outputGlbPath + " -d", cwd=newCwd, shell=True)
-			print("obj2gltf: Draco compression finished")
+			success = subprocess.run("gltf-pipeline", "-i", outputGlbPath, "-o", outputGlbPath, "-d", cwd=newCwd)
```
Would be a good idea to add a check for the program's execution status

UdomkarnBoonyaprasert replied:
Check for return code added

boonwj commented on HoloPipelines/test/test_pipelineController.py:
```diff
@@ -6,45 +6,29 @@
+	output = subprocess.run(["python", "pipelineController.py", "-c", "test/testList.json", "s0", "-p", "1"], cwd=newCwd)
+	assert 1 == output.returncode
 
 def test_main_s0_2():
```
Test name is test_main_s0_2, but s0 does not appear anywhere. Test function names can be more explicit to indicate what you are testing for. e.g. missing pipeline etc.

UdomkarnBoonyaprasert replied:
Renamed to test_main_missing_pipeline

boonwj commented on HoloPipelines/test/test_pipelineController.py:
```diff
@@ -6,45 +6,29 @@
-	assert "1" in str(output)
-	'''output = subprocess.check_output('python pipelineController.py -c test/testList.json s3 -p 0', cwd=newCwd, shell=True).decode("utf-8")
-	assert "pipelineController: no pipeline with such ID" in str(output)'''
+	output = subprocess.run(["python", "pipelineController.py", "-c", "test/testList.json", "s3", "-p", "0"], cwd=newCwd)
```
same comment as above.

UdomkarnBoonyaprasert replied:
Renamed to test_main_s0_missing_pipeline_with_extra_arguments

boonwj commented on HoloPipelines/test/test_pipelineController.py:
```diff
@@ -6,45 +6,29 @@
+	output = subprocess.run(["python", "pipelineController.py", "-c", "test/testList.json", "s0"], cwd=newCwd)
+	assert 0 == output.returncode
 
 def test_main_s0_1():
```
Is s0_1's main test for the parameters option? Perhaps name it that way.

UdomkarnBoonyaprasert replied:
Renamed to test_main_s0_invalid_number_of_argument

boonwj commented on HoloPipelines/test/test_pipelineController.py:
```diff
@@ -6,45 +6,29 @@
+	output = subprocess.run(["python", "pipelineController.py", "-c", "test/testList.json", "--info", "dicom2glb"], cwd=newCwd)
+	assert 0 == output.returncode
 
 def test_main_info_1():
```
missing info testing can be more explicit in name

UdomkarnBoonyaprasert replied:
Renamed to test_main_info_for_missing_pipeline

boonwj commented on HoloPipelines/README.md:
````diff
@@ -18,14 +18,12 @@ sudo apt-get install libcdcm-tools
-or
-```
-gdcmconv
+PATH=$PATH:pathToYourLocalGDCMbin
````
Do they type this command as it is? Or added to a file?

UdomkarnBoonyaprasert replied:
readme updated

boonwj reviewed (changes requested):
Looking good. On the whole the code looks much better now.

Some small changes requested for error handling and test naming confusion.

UdomkarnBoonyaprasert committed:
PIPELINE: PR clean up
-test cases naming
-returncode checking in compObj2glbWrapper.py

UdomkarnBoonyaprasert committed:
PIPELINE: PR cleanup
delete some lines

UdomkarnBoonyaprasert committed:
PIPELINE: ran code through black

UdomkarnBoonyaprasert committed:
PIPELINE: PR issue solving
solving issues raise by CodeFactor

boonwj commented on HoloPipelines/README.md:
````diff
@@ -19,10 +19,10 @@ sudo apt-get install libcdcm-tools
 
 ```
-PATH=$PATH:pathToYourLocalGDCMbin
+PATH=$PATH:"directory_to_GDCM_bin_folder"
````
I am running through the installation and i realised that directory to GDCM bin folder will not be known when you install via package manager. This was a legacy installation instruction when you had to compile your own binary right?

boonwj replied:
Removed it.

boonwj commented on HoloPipelines/README.md:
````diff
@@ -1,41 +1,87 @@
+npm install -g gltf-pipeline
+```
+
+(optional)Run *setupExample.py* to download sample dicom and nifti files which can be used for manual testing.
````
Oh this README step is no longer needed since your test code performs the download

boonwj replied:
Updated the README.md

boonwj commented on HoloPipelines/README.md:
```diff
@@ -1,41 +1,87 @@
+
+(optional)Run *setupExample.py* to download sample dicom and nifti files which can be used for manual testing.
+
+Downloaded files from *setupExample.py* and all the produced 3D models can be deleted by running *cleanUp.py*
```
Remove setupExample.py reference.

boonwj replied:
Removed.

boonwj commented on HoloPipelines/pipelineController.py:
```diff
@@ -0,0 +1,96 @@
+parser.add_argument(
+    "-c",
+    "--config",
+    default="pipelines/pipelineList.json",
```
Configuration file should be in a higher level directory, ./pipelineList.json

boonwj replied:
Changes made.

boonwj committed:
PIPELINE: Update README.md

- Remove obselete reference to setupExample.py
- Remove PATH instruction which was leftover from the steps to build
gdcm manually
- Shifted cleanup.py command to testing section

boonwj committed:
PIPELINE: Shift pipeline config to base dir

- Move configuration file to base directory for visibility
- Change default configfile parameter of reflect new location

boonwj committed:
PIPELINE: Update INFO metavar and README

Change -i metavar from INFO to NAME, to better reflect its
functionality.

Update README with new usage and rephrase testing section

boonwj commented on HoloPipelines/README.md:
```diff
@@ -1,41 +1,87 @@
+                        path to pipeline config file relative to
+                        pipelineController
+  -l, --ls              list all the available piplines
+  -i INFO, --info INFO  get info from pipeline's name
```
NAME will be a better metavariable than INFO, as you are looking for info on pipeline name.

boonwj replied:
Changed to NAME metavar.

boonwj reviewed (changes requested):
I ran the code through flake8 and it caught a bunch of errors that should be resolved. Mainly due to unused imports.

```
HoloPipelines/components/lungSegment/utils.py:114:32: E203 whitespace before ':'
HoloPipelines/components/lungSegment/utils.py:117:42: E203 whitespace before ':'
HoloPipelines/components/lungSegment/utils.py:129:31: E203 whitespace before ':'
HoloPipelines/components/lungSegment/utils.py:129:60: E203 whitespace before ':'
HoloPipelines/components/lungSegment/utils.py:163:32: E203 whitespace before ':'
HoloPipelines/components/lungSegment/utils.py:164:32: E203 whitespace before ':'
HoloPipelines/components/lungSegment/utils.py:165:14: E203 whitespace before ':'
HoloPipelines/components/lungSegment/utils.py:169:32: E203 whitespace before ':'
HoloPipelines/components/lungSegment/utils.py:170:32: E203 whitespace before ':'
HoloPipelines/components/lungSegment/utils.py:171:31: E203 whitespace before ':'
HoloPipelines/components/compNumpy2obj.py:3:1: F401 'pydicom as dicom' imported but unused
HoloPipelines/components/compNumpy2obj.py:4:1: F401 'pydicom.pixel_data_handlers.gdcm_handler' imported but unused
HoloPipelines/components/compNumpy2obj.py:5:1: F401 'os' imported but unused
HoloPipelines/components/compNumpy2obj.py:6:1: F401 'matplotlib.pyplot as plt' imported but unused
HoloPipelines/components/compNumpy2obj.py:7:1: F401 'glob.glob' imported but unused
HoloPipelines/components/compNumpy2obj.py:8:1: F401 'mpl_toolkits.mplot3d.art3d.Poly3DCollection' imported but unused
HoloPipelines/components/compNumpy2obj.py:9:1: F401 'scipy.ndimage' imported but unused
HoloPipelines/components/compNumpy2obj.py:10:1: F401 'skimage.morphology' imported but unused
HoloPipelines/components/compNumpy2obj.py:12:1: F401 'skimage.transform.resize' imported but unused
HoloPipelines/components/compNumpy2obj.py:13:1: F401 'sklearn.cluster.KMeans' imported but unused
HoloPipelines/components/compNumpy2obj.py:42:9: F841 local variable 'e' is assigned to but never used
HoloPipelines/components/compDcm2nifti.py:2:1: F401 'pathlib' imported but unused
HoloPipelines/components/compNifti2numpy.py:3:1: F401 'nilearn.image.resample_img' imported but unused
HoloPipelines/components/compNifti2numpy.py:4:1: F401 'scipy' imported but unused
HoloPipelines/components/compNifti2numpy.py:5:1: F401 'pathlib' imported but unused
HoloPipelines/components/compNifti2numpy.py:6:1: F401 'platform' imported but unused
HoloPipelines/components/compNifti2numpy.py:12:5: F841 local variable 'header' is assigned to but never used
HoloPipelines/components/lungSegment/main.py:1:1: F401 'numpy as np' imported but unused
HoloPipelines/components/lungSegment/main.py:33:5: E741 ambiguous variable name 'I'
HoloPipelines/components/lungSegment/main.py:35:5: E741 ambiguous variable name 'I'
HoloPipelines/components/lungSegment/segment_lung.py:41:30: E203 whitespace before ':'
HoloPipelines/components/lungSegment/segment_lung.py:42:30: E203 whitespace before ':'
HoloPipelines/components/lungSegment/segment_airway.py:8:1: F841 local variable 'e' is assigned to but never used
HoloPipelines/components/compDicom2numpy.py:5:1: F401 'glob.glob' imported but unused
HoloPipelines/components/compDicom2numpy.py:6:1: F401 'mpl_toolkits.mplot3d.art3d.Poly3DCollection' imported but unused
HoloPipelines/components/compDicom2numpy.py:9:1: F401 'time' imported but unused
HoloPipelines/components/compDicom2numpy.py:22:5: F841 local variable 'e' is assigned to but never used
HoloPipelines/components/compDicom2numpy.py:63:5: F841 local variable 'e' is assigned to but never used
HoloPipelines/components/compDicom2numpy.py:69:9: F821 undefined name 'sys'
```

boonwj committed:
PIPELINE: fix unnecessary and missing imports

Remove every unused import
Add missing imports needed in the project

boonwj commented on HoloPipelines/pipelines/components/compNifti2numpy.py:
```diff
@@ -0,0 +1,43 @@
+def resample(dataPath, new_spacing=[1, 1, 1]):
+    image = dataPath
+    originalShape = image.shape[:3]
+    header = image.header
```
local variable 'header' is assigned to but never used

UdomkarnBoonyaprasert replied:
`header = image.header` deleted

boonwj commented on HoloPipelines/pipelines/components/lungSegment/segment_airway.py:
```diff
@@ -0,0 +1,59 @@
+
+try:
+    from utils import close_space_dilation, generate_structure_trachea, generate_initLoc
+except Exception as e:
```
local variable 'e' is assigned to but never used. Log the exception that was caught.

UdomkarnBoonyaprasert replied:
'e' now used in logging

boonwj commented on HoloPipelines/pipelines/pipelineDicom2glb.py:
```diff
@@ -0,0 +1,27 @@
+        )
+        + ".obj",
+    )
+    generatedGlbPath = compObj2glbWrapper.main(
```
local variable 'generatedGlbPath' is assigned to but never used.

Print the path to find the new output?

UdomkarnBoonyaprasert replied:
output path variable now used in final print of a pipeline

boonwj commented on HoloPipelines/pipelines/pipelineLungDicom2glb.py:
```diff
@@ -0,0 +1,41 @@
+        0.5,
+        str(pathlib.Path.cwd().joinpath("output", "OBJ", "nifti2glb_tempObj.obj")),
+    )
+    generatedGlbPath = compObj2glbWrapper.main(
```
local variable 'generatedGlbPath' is assigned to but never used. 

The output path of the glb can be printed.

UdomkarnBoonyaprasert replied:
output path variable now used in final print of a pipeline

boonwj commented on HoloPipelines/pipelines/pipelineNifti2glb.py:
```diff
@@ -0,0 +1,25 @@
+        threshold,
+        str(pathlib.Path.cwd().joinpath("output", "OBJ", "nifti2glb_tempObj.obj")),
+    )
+    generatedGlbPath = compObj2glbWrapper.main(
```
Same here, unused variable. Make use of it.

UdomkarnBoonyaprasert replied:
output path variable now used in final print of a pipeline

boonwj commented on HoloPipelines/pipelines/pipelineNifti2obj.py:
```diff
@@ -0,0 +1,17 @@
+
+def main(inputNiftiPath, outputObjPath, threshold, flipNpy=False):
+    generatedNumpyList = compNifti2numpy.main(str(pathlib.Path(inputNiftiPath)))
+    generatedObjPath = compNumpy2obj.main(
```
same here. make use of this.

UdomkarnBoonyaprasert replied:
output path variable now used in final print of a pipeline

boonwj reviewed (changes requested):
Cleaned up most of the import errors. But I am leaving the following for you to handle.

```
./pipelines/pipelineNifti2obj.py:10:5: F841 local variable 'generatedObjPath' is assigned to but never used
./pipelines/pipelineDicom2glb.py:20:5: F841 local variable 'generatedGlbPath' is assigned to but never used
./pipelines/pipelineNifti2glb.py:15:5: F841 local variable 'generatedGlbPath' is assigned to but never used
./pipelines/pipelineLungDicom2glb.py:31:5: F841 local variable 'generatedGlbPath' is assigned to but never used
./pipelines/components/compNifti2numpy.py:8:5: F841 local variable 'header' is assigned to but never used
./pipelines/components/lungSegment/segment_airway.py:8:1: F841 local variable 'e' is assigned to but never used
```

nielsboecker commented on HoloPipelines/test/testList.json:
```diff
@@ -0,0 +1,47 @@
+{
```
This file should have a better name, yes it is a list but what does it contain?

UdomkarnBoonyaprasert replied:
Renamed to 'pipelineListForTesting.json'

nielsboecker commented on HoloPipelines/test/sample0.py:
```diff
@@ -0,0 +1 @@
+print("Hello World! from 0")
```
Hmm does this really make sense? Are you just testing the Python `print` function here? I am not convinced if it makes sense. Even if yes, the file should have a better name. You want it to represent a dummy pipeline? call it `pipelineHelloWorld` or whatever. But let's first discuss why this is needed.

UdomkarnBoonyaprasert replied:
'sample0.py' renamed to 'pipelineHelloWorld.py'. And yes this script is just a dummy pipeline.

nielsboecker commented on HoloPipelines/test/testList.json:
```diff
@@ -0,0 +1,47 @@
+      {
+         "name":"sample0",
+         "src":"test/sample0.py",
+         "info":"A pipline for testing. It will printout 'Hello World! from 0'",
```
Typo: `pipeline`

UdomkarnBoonyaprasert replied:
corrected

nielsboecker commented on HoloPipelines/test/test_pipelines.py:
```diff
@@ -0,0 +1,144 @@
+niftiPath = thisCwd.joinpath("medicalScans", "nifti")
+
+outputPath = pathlib.Path.cwd().joinpath("output")
+objPath = outputPath.joinpath("OBJ")
```
Minor issue, but why are the output paths CAPITALISED while the input paths are small letters?

nielsboecker replied:
Also I noticed that several input / output paths seem to get assembled multiple times throught different parts of the code. Would it maybe make sense to have one file where the paths are joined once and for all, and other files just import that?

UdomkarnBoonyaprasert replied:
OBJ and GLB changed to lower case. I've added a new components called 'compCommonPath.py' which is used to store frequently used file path (dicom, nifti, obj and glb)

nielsboecker commented on HoloPipelines/test/test_pipelines.py:
```diff
@@ -0,0 +1,144 @@
+        print("Beginning dicom sample download...")
+
+        url = "https://holoblob.blob.core.windows.net/test/3_Axial_CE.zip"
+        urllib.request.urlretrieve(url, str(thisCwd.joinpath("__temp__.zip")))
```
use variable for `__temp__.zip`

UdomkarnBoonyaprasert replied:
changed to ```zipFileName = "__temp__.zip"```

nielsboecker commented on HoloPipelines/cleanUp.py:
```diff
@@ -0,0 +1,30 @@
+import os
+import shutil
+
+if not os.path.exists("medicalScans"):
```
Some issues here:
1. the if / else conditionals in this file contain much duplicate code. If you really need if/else, just do the parts that differ in the conditional section, and have the shared code run in all cases.
1. Why does a cleanUp file _create_ directories? If you want to make sure the directories exist, you should probably check/create them in the setUp parts of testing, where you also do the downloads. Otherwise you don't know if someone ran the cleanUp script before (which they won't, in case of the first run)
1. In fact i just saw that you create the directories in `pipelineController.main` anyway, so no need to do it here. Just delete them I guess.

UdomkarnBoonyaprasert replied:
'cleanUp.py' removed as clean up is done within the fixture in test_pipelines.py

nielsboecker commented on HoloPipelines/cleanUp.py:
```diff
@@ -0,0 +1,30 @@
+import os
```
Could be renamed to `cleanUpTestInputData.py`. Also this should be in the `test` directory as it really relates to testing. When you move it please remember to update any paths if needed.

UdomkarnBoonyaprasert replied:
'cleanUp.py' removed as clean up is done within the fixture in test_pipelines.py

nielsboecker commented on HoloPipelines/pipelineController.py:
```diff
@@ -0,0 +1,101 @@
+
+newCwd = str(pathlib.Path(str(os.path.dirname(os.path.realpath(__file__)))))
+
+parser = argparse.ArgumentParser(description="Selct pipeline to process")
```
Typo `Select`

UdomkarnBoonyaprasert replied:
Corrected

nielsboecker commented on HoloPipelines/pipelineController.py:
```diff
@@ -0,0 +1,101 @@
+    help="path to pipeline config file relative to pipelineController",
+)
+parser.add_argument(
+    "-l", "--ls", action="store_true", help="list all the available piplines"
```
Typo `pipelines`
(Hint: using a full-fledged IDE like PyCharm makes life easier because it will point out such typos)

UdomkarnBoonyaprasert replied:
Corrected. Yeah, I should look into using an actual IDE

UdomkarnBoonyaprasert committed:
PIPELINE: make use of unsed variables
Removed variable header in compNifti2numpy.py and print out the return value of pipelines(output path)

UdomkarnBoonyaprasert committed:
PIPELINE: ran code through Black

nielsboecker commented on HoloPipelines/pipelineList.json:
```diff
@@ -0,0 +1,38 @@
+{
+   "p0":
```
Can I ask you for another favour? Could you please use expressive ids? In fact, please just take the fields you use for name, like `nifti2obj`, and then use human-readable names for the `name` field, like "NIfTI to DICOM Conversion". Please remember that this will also mean a few other files need to be changed, for instance the tests.

UdomkarnBoonyaprasert replied:
All IDs renamed to the values from the 'name' field. And name has been updated with a more readable value.

nielsboecker commented on HoloPipelines/pipelines/components/compDicom2numpy.py:
```diff
@@ -0,0 +1,89 @@
+import numpy as np
+import pydicom as dicom
+import pydicom.pixel_data_handlers.gdcm_handler
```
unused import

UdomkarnBoonyaprasert replied:
Removed

nielsboecker commented on HoloPipelines/pipelines/components/lungSegment/LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
```
Is the complete content of this directory copied from this person's code? Or did you take and modify it significantly? Anyways I feel like we should credit it more clearly, and include a link to their repo (maybe in the README)

UdomkarnBoonyaprasert replied:
Credit added in the readme.md (including the link to the original repository)

UdomkarnBoonyaprasert committed:
PIPELINE: update file naming
-delete cleanUp.py as clean up is already done as part of fixture in testing
-add compCommonPath.py for the frequently used file path in pipelines and components
-update pipeline IDs and names in pipeline config files
-update test case to match the new config file
-minor changes due to typos

UdomkarnBoonyaprasert committed:
PIPELINE: ran code through black

UdomkarnBoonyaprasert committed:
PIPELINE: update credit in readme
Added credit for the lung segmentation (Jie Yang) in the readme.md

boonwj commented:
I think the new changes and developments broke some stuff.

1. README now conflicts with what was done in the #15. Will definitely have to resolve the changes between the 2 files
2. Tests now fail because of the newly introduced commonPath component. I am not entirely convinced that it qualifies as a "component", to me it feels more like a configuration instead. But it will do for now.

UdomkarnBoonyaprasert committed:
PIPELINE: fix output path

boonwj commented on HoloPipelines/pipelineController.py:
```diff
@@ -45,8 +45,8 @@ def main():
-        os.mkdir("output/OBJ")
-        os.mkdir("output/GLB")
+        os.mkdir("output/obj")
+        os.mkdir("output/glb")
```
It's does not look right for paths to be hardcoded here when your other components take reference elsewhere. It would be better if you use the same reference everywhere it is needed.

UdomkarnBoonyaprasert replied:
Changed to compCommonPath

boonwj reviewed (changes requested):
Don't use hardcoded filepaths in pipelineController if the other components take reference from somewhere else.

UdomkarnBoonyaprasert committed:
Merge branch 'dev' into PIPELINE/controller

UdomkarnBoonyaprasert committed:
PIPELINE: update pipelineController to use compCommonPath

nielsboecker commented on HoloPipelines/README.md:
```diff
@@ -111,8 +115,28 @@ Execute tests by running the following command in `HoloPipelines` directory:
+
+
+## API specification
+> TODO: update this
```
TODO is resolved

UdomkarnBoonyaprasert committed:
PIPELINE: remove comment from readme

UdomkarnBoonyaprasert committed:
PIPELINE: fix mkdir in commonPath

UdomkarnBoonyaprasert committed:
PIPELINE: ran code through black

boonwj reviewed (approved):
I think I am ok with the state of the PR now. @nbckr: For final comments.

nbckr commented:
Looks good!

nielsboecker reviewed (approved):
Looks good!
